src/firefly/java/edu/caltech/ipac/firefly/server/visualize/hipslists/HiPSMasterList.java

```diff
@@ -0,0 +1,210 @@
+                    orderedSources.add(defaultSourceOrder[i]);
+                }
+            }
+        } else {
+            // add element defined in sort order first
+            for (int i = 0; i < workingOrder.length; i++) {
+                if (Arrays.asList(workingSources).contains(workingOrder[i])) {
+                    orderedSources.add(workingOrder[i]);
+                }
+            }
+            // add elements defined in sources but not defined in sort order
+            for (int i = 0; i < defaultSourceOrder.length; i++) {
+                if (!orderedSources.contains(defaultSourceOrder[i]) &&
+                        Arrays.asList(workingSources).contains(defaultSourceOrder[i])) {
+                    orderedSources.add(defaultSourceOrder[i]);
+                }
+            }
+        }
+
+        try {
+            for (String source : orderedSources) {
+                HiPSMasterListSourceType hipsls = sources.get(source);
+
+                if (hipsls != null) {
+                    List<HiPSMasterListEntry> hipsL = hipsls.getHiPSListData(workingTypes, source);
+                    if (hipsL != null) allSourceData.addAll(hipsL);
+                }
+            }
+
+            if (allSourceData.size() == 0) {
+                throw new IOException("[HiPS_MASTER]: no HiPS found");
+            }
+
+            DataGroup dg = createTableDataFromListEntry(allSourceData);
+            dg.shrinkToFitData();
+
+            setupMeta(dg, request);
+
+            FileInfo finfo = EmbeddedDbUtil.ingestDataGroup(dbFile, dg, dbAdapter, "data");
+            return finfo;
+        } catch (Exception e) {
+            throw new DataAccessException("[HiPS_MASTER]: Unable to get HiPS");
+        }
+    }
+
+    private void setupMeta(DataGroup dg, ServerRequest req) {
+        Map<String, DataGroup.Attribute> cmeta = dg.getAttributes();
+        TableMeta meta = new TableMeta();
+
+
+        SearchProcessor.prepareTableMetaDef(meta, Arrays.asList(dg.getDataDefinitions()), req);
+        for (String key : meta.getAttributes().keySet()) {
+            if (!cmeta.containsKey(key)) {
+                dg.addAttribute(key, meta.getAttribute(key));
+            }
+        }
+
+        int    sWidth = 30;
+
+        for (PARAMS oneCol : PARAMS.values()) {
+            String colName = oneCol.getKey();
+
+            dg.addAttribute(DataSetParser.makeAttribKey(DataSetParser.LABEL_TAG, colName),
+                                                        oneCol.getTitle());
+            if (oneCol.getMetaClass() != String.class) continue;
+
+            DataType colDT = dg.getDataDefintion(colName);
+            if (colDT != null) {
+                if (colDT.getFormatInfo().getWidth() > sWidth) {
+                    dg.addAttribute(DataSetParser.makeAttribKey(DataSetParser.WIDTH_TAG, colName),
+                                    Integer.toString(sWidth));
+                }
```

Comment:
shrinkToFitData() will update the column width to become to be  the maximum among all strings of that column. Is this the way you mentioned that we can use to reduce the column width in case the width number is too high for the UI?  Please tell the fixing method.

Reply:
You are right.  I read it wrong.  I though `sWidth` is the current column's width, and not a constant 30.  That's why I thought the logic was awkward.  Ignore my comment.